android/src/main/java/io/sentry/RNSentryModule.java

```diff
@@ -190,4 +219,114 @@ private Level logLevel(int level) {
                 return Level.OFF;
         }
     }
+
+    @ReactMethod
+    public void setUser(final ReadableMap user) {
+        if (user == null) {
+            Sentry.setUser(null);
+        } else {
+            User userInstance = new User();
+
+            // only support these defined keys for now.
+            if (user.hasKey("email")) {
+                userInstance.setEmail(user.getString("email"));
+            } else {
+                userInstance.setEmail(null);
+            }
+
+            if (user.hasKey("id")) {
+                userInstance.setId(user.getString("id"));
+            } else {
+                userInstance.setId(null);
+            }
+
+            if (user.hasKey("username")) {
+                userInstance.setUsername(user.getString("username"));
+            } else {
+                userInstance.setUsername(null);
+            }
+
+            if (user.hasKey("ip_address")) {
+                userInstance.setIpAddress(user.getString("ip_address"));
+            } else {
+                userInstance.setIpAddress(null);
+            }
+
+            Sentry.setUser(userInstance);
+        }
+    }
+
+    @ReactMethod
+    public void addBreadcrumb(final ReadableMap breadcrumb) {
+        Breadcrumb breadcrumbInstance = new Breadcrumb();
+
+        if (breadcrumb.hasKey("message")) {
+            breadcrumbInstance.setMessage(breadcrumb.getString("message"));
+        }
+
+        if (breadcrumb.hasKey("type")) {
+            breadcrumbInstance.setType(breadcrumb.getString("type"));
+        }
+
+        if (breadcrumb.hasKey("category")) {
+            breadcrumbInstance.setCategory(breadcrumb.getString("category"));
+        }
+
+        if (breadcrumb.hasKey("level")) {
+            switch (breadcrumb.getString("level")) {
+                case "fatal":
+                    breadcrumbInstance.setLevel(SentryLevel.FATAL);
+                    break;
+                case "warning":
+                    breadcrumbInstance.setLevel(SentryLevel.WARNING);
+                    break;
+                case "info":
+                    breadcrumbInstance.setLevel(SentryLevel.INFO);
+                    break;
+                case "debug":
+                    breadcrumbInstance.setLevel(SentryLevel.DEBUG);
+                    break;
+                case "error":
+                    breadcrumbInstance.setLevel(SentryLevel.ERROR);
+                    break;
+                default:
+                    breadcrumbInstance.setLevel(SentryLevel.ERROR);
+                    break;
+            }
+        }
+
+        if (breadcrumb.hasKey("data")) {
+            ReadableMap data = breadcrumb.getMap("data");
+            ReadableMapKeySetIterator it = data.keySetIterator();
+            while (it.hasNextKey()) {
+                String key = it.nextKey();
+                String value = data.getString(key);
```

Comment:
We can't assume this is `String`. We need to test the type somehow or try/catch this straight to string and log out the errors.
Users can add complex objects as values or pretty much anything. We could extend this to test for `Serializable` or other types in the future but for now just please gracefully handle the fact it might not be `String`

Reply:
All first-level values in data that aren't strings are actually passed through `JSON.stringify` in the `wrapper.ts` so they would be strings before they get here. I can do the same for all custom values passed to user then.

```js
const stringifiedData: { [key: string]: any } = {};
    if (typeof breadcrumb.data !== "undefined") {
      Object.keys(breadcrumb.data).forEach((dataKey) => {
        if (typeof breadcrumb.data !== "undefined") {
          const value = breadcrumb.data[dataKey];
          stringifiedData[dataKey] =
            typeof value === "string" ? value : JSON.stringify(value);
        }
      });
    }

    // tslint:disable-next-line: no-unsafe-any
    return RNSentry.addBreadcrumb({
      ...breadcrumb,
      data: stringifiedData
    });
```